src/crimson/os/seastore/journal/circular_bounded_journal.h

```diff
@@ -0,0 +1,320 @@
+
+  struct cbj_header_t;
+  using write_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::erange>;
+  /*
+   * append_record
+   *
+   * append data to current write position of CircularBoundedJournal
+   *
+   * @param bufferlist to write
+   * @param rbm_abs_addr where data is written
+   *
+   */
+  write_ertr::future<> append_record(ceph::bufferlist bl, rbm_abs_addr addr);
+  /*
+   * device_write_bl
+   *
+   * @param device address to write
+   * @param bufferlist to write
+   *
+   */
+  write_ertr::future<> device_write_bl(rbm_abs_addr offset, ceph::bufferlist &bl);
+
+  using read_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg,
+    crimson::ct_error::enoent,
+    crimson::ct_error::erange>;
+  using read_record_ertr = read_ertr;
+  using read_record_ret = read_record_ertr::future<
+	std::optional<std::pair<record_group_header_t, bufferlist>>
+	>;
+  using read_header_ertr = read_ertr;
+  using read_header_ret = read_header_ertr::future<
+	std::optional<std::pair<cbj_header_t, bufferlist>>
+	>;
+  /*
+   * read_record
+   *
+   * read record from given address
+   *
+   * @param paddr_t to read
+   *
+   */
+  read_record_ret read_record(paddr_t offset);
+  /*
+   * read_header
+   *
+   * read header block from given absolute address
+   *
+   * @param absolute address
+   *
+   */
+  read_header_ret read_header(rbm_abs_addr start);
+
+  ceph::bufferlist encode_header();
+
+  using mkfs_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg
+  >;
+  using mkfs_ret = mkfs_ertr::future<>;
+
+  /*
+   * mkfs
+   *
+   * make a new journal layout even if old journal exists
+   *
+   * @param mkfs_config_t
+   *
+   */
+  mkfs_ret mkfs(const mkfs_config_t& config);
+
+
+  /**
+   * CircularBoundedJournal structure
+   *
+   * +-------------------------------------------------------+
+   * |   header    | record | record | record | record | ... |
+   * +-------------------------------------------------------+
+   *               ^-----------block aligned-----------------^
+   * <----fixed---->
+   */
+
+
+  /**
+   *
+   * CircularBoundedJournal write
+   *
+   * NVMe will support a large block write (< 512KB) with atomic write unit command.
+   * With this command, we expect that the most of incoming data can be stored
+   * as a single write call, which has lower overhead than existing
+   * way that uses a combination of system calls such as write() and sync().
+   *
+   */
+
+  struct cbj_header_t {
+    uint64_t magic = CBJOURNAL_MAGIC;
+    uuid_d uuid;
+    uint64_t block_size = 0; // block size of underlying device
+    uint64_t size = 0;   // max length of journal
+
+    // start offset of CircularBoundedJournal in the device
+    rbm_abs_addr journal_tail = 0;
+    // start address where the newest record will be written
+    rbm_abs_addr written_to = 0;
```

Comment:
(sorry, I left this comment yesterday, but it doesn't seem to show up on the main page)

As I mentioned before, this field will, in general, be out of date since you do not write the header out on commit. For this reason, leaving it in the header is dangerous as someone might write code that assumes that the header encoded on disk has a valid written_to value. replay() immediately overwrites this value after calling open_for_write() -- there's no obvious reason to write it into the header at all. Make this a member of CircularBoundedJournal directly.